### ISSUE TITLE
train while use lora and bugs

### ISSUE THREAD

**Issue by MapleMaster1023** [verl-project/verl#6380]
### System Info

I've trained a sft-trained Qwen3-8b-sft model with GRPO in verl however I met 2 bugs
1:when I get to the last steps,the trainer exit unexpectedly but the second to last seems to be right
2:when i use your python scripts to transfer pt.s to safetensors,here come out an adapter folder and a model folder
The model in model folder show zero different with the base one，but in the train，i deed see reward raise up.
so i check the adapter_config.json in adapter folder.
It ignored the LoRA rank and alpha specified in my training script and arbitrarily reset them to zero, so I had to modify the JSON and manually merge again. But the problem is, with so many bugs throughout the process, I can't verify whether this final merged model is actually the penultimate model from the training.

here is my train.sh
`#!/bin/bash

export REWARD_SILENT=1
export REWARD_DEBUG_FILE="/userhome/reward_debug.jsonl"

export REWARD_DEBUG_SAMPLES=10

export TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_cache
export TRITON_CACHE_DIR=/tmp/triton_cache
export TMPDIR=/tmp
export TEMP=/tmp
export TMP=/tmp

mkdir -p /tmp/torchinductor_cache /tmp/triton_cache
export TORCH_COMPILE_DISABLE=1

CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
MODEL_PATH=/userhome/outputs/qwen3-8b-dora64-REA-GEN-merged
TRAIN_FILE=/userhome/bioprobench_grpo.jsonl
TEST_FILE=/userhome/REA-GEN_test_outputs——QWEN3/grpo_test_64_fixed.jsonl
export WANDB_API_KEY="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"

NGPUS_PER_NODE=8
NNODES=1
TRAIN_ROLLOUT_L0G=/userhome/grpo-gen-new
TRAIN_BATCH_SIZE=128
PPO_MINI_BATCH_SIZE=64
PPO_MICRO_BATCH_SIZE_PER_GPU=2
LOG_PROB_MICRO_BATCH_SIZE_PER_GPU=8
MAX_PROMPT_LENGTH=4096
MAX_RESPONSE_LENGTH=4096
PPO_MAX_TOKEN_LEN_PER_GPU=8192

ACTOR_LR=1e-5
ROLLOUT_TP=1
ROLLOUT_N=4
ROLLOUT_GPU_MEM_UTIL=0.65

REWARD_NUM_WORKERS=2

PROJECT_NAME=verl_bioprobench_grpo1e-5
EXPERIMENT_NAME=qwen3_8b_dora64_grpo_biopro_new
SAVE_FREQ=50
TEST_FREQ=20
TOTAL_EPOCHS=3

DATA=(
    algorithm.adv_estimator=grpo
    algorithm.use_kl_in_reward=False
    data.train_files="['${TRAIN_FILE}']"
    data.val_files="['${TEST_FILE}']"
    data.prompt_key=prompt
    data.return_raw_chat=False
    data.train_batch_size=${TRAIN_BATCH_SIZE}
    data.max_prompt_length=${MAX_PROMPT_LENGTH}
    data.max_response_length=${MAX_RESPONSE_LENGTH}
    data.filter_overlong_prompts=True
    data.truncation='error'
    data.shuffle=False
    data.seed=2026
    
)

MODEL=(
    actor_rollout_ref.model.path=${MODEL_PATH}
    actor_rollout_ref.model.use_remove_padding=True
    actor_rollout_ref.model.enable_gradient_checkpointing=True
    +actor_rollout_ref.model.override_config="{attn_implementation:sdpa}"
    actor_rollout_ref.model.lora_rank=64
    actor_rollout_ref.model.lora_alpha=128
    actor_rollout_ref.model.target_modules=all-linear
    actor_rollout_ref.model.lora.merge=True
)

ACTOR=(
    actor_rollout_ref.actor.optim.lr=${ACTOR_LR}
    actor_rollout_ref.actor.ppo_mini_batch_size=${PPO_MINI_BATCH_SIZE}
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=${PPO_MICRO_BATCH_SIZE_PER_GPU}
    actor_rollout_ref.actor.use_dynamic_bsz=True
    actor_rollout_ref.actor.ppo_max_token_len_per_gpu=${PPO_MAX_TOKEN_LEN_PER_GPU}
    actor_rollout_ref.actor.use_kl_loss=True
    actor_rollout_ref.actor.kl_loss_coef=0.001
    actor_rollout_ref.actor.kl_loss_type=low_var_kl
    actor_rollout_ref.actor.fsdp_config.model_dtype=bfloat16
    ++actor_rollout_ref.rollout.load_format=safetensors

)

ROLLOUT=(
    actor_rollout_ref.rollout.name=vllm
    actor_rollout_ref.rollout.tensor_model_parallel_size=${ROLLOUT_TP}
    actor_rollout_ref.rollout.gpu_memory_utilization=${ROLLOUT_GPU_MEM_UTIL}
    actor_rollout_ref.rollout.n=${ROLLOUT_N}
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=${LOG_PROB_MICRO_BATCH_SIZE_PER_GPU}
    actor_rollout_ref.rollout.dtype=bfloat16
    ++actor_rollout_ref.rollout.enforce_eager=True
)

REWARD=(
    reward.num_workers=${REWARD_NUM_WORKERS}
    reward.reward_model.enable=False
    reward.custom_reward_function.path=/userhome/verl/examples/grpo_trainer/reward_with_server.py
    reward.custom_reward_function.name=compute_score
    +reward.custom_reward_function.reward_kwargs.embedding_model_dir=/userhome/all-mpnet-base-v2
)

TRAINER=(
    trainer.balance_batch=True
    trainer.logger='["console","wandb"]'
    trainer.project_name=${PROJECT_NAME}
    trainer.experiment_name=${EXPERIMENT_NAME}
    trainer.n_gpus_per_node=${NGPUS_PER_NODE}
    trainer.nnodes=${NNODES}
    trainer.save_freq=${SAVE_FREQ}
    trainer.test_freq=${TEST_FREQ}
    trainer.total_epochs=${TOTAL_EPOCHS}
    trainer.rollout_data_dir=${TRAIN_ROLLOUT_L0G}
    trainer.validation_data_dir=${TEST_ROLLOUT_L0G}
    +trainer.keep_last_n_checkpoints=3 
    trainer.resume_mode=auto              
    trainer.resume_from_path=null 
    trainer.default_local_dir=/userhome/grpo-gen-new
)

python3 -m verl.trainer.main_ppo \
    "${DATA[@]}" \
    "${MODEL[@]}" \
    "${ACTOR[@]}" \
    "${ROLLOUT[@]}" \
    "${REWARD[@]}" \
    "${TRAINER[@]}" \
    "$@"
`
and here is the adapter_config.json in out put(maybe your pt. to safetensors scripts dont meet the lora rank and alpha)
`{
    "task_type": null,
    "peft_type": "LORA",
    "auto_mapping": null,
    "peft_version": "0.19.1",
    "base_model_name_or_path": null,
    "revision": null,
    "inference_mode": false,
    "r": 64,
    "target_modules": [
        "down_proj",
        "up_proj",
        "k_proj",
        "gate_proj",
        "v_proj",
        "o_proj",
        "q_proj"
    ],
    "exclude_modules": null,
    "lora_alpha": 0,
    "lora_dropout": 0.0,
    "fan_in_fan_out": false,
    "bias": "none",
    "use_rslora": false,
    "modules_to_save": null,
    "init_lora_weights": true,
    "layers_to_transform": null,
    "layers_pattern": null,
    "rank_pattern": {},
    "alpha_pattern": {},
    "megatron_config": null,
    "megatron_core": "megatron.core",
    "trainable_token_indices": null,
    "loftq_config": {},
    "eva_config": null,
    "corda_config": null,
    "lora_ga_config": null,
    "use_dora": false,
    "alora_invocation_tokens": null,
    "use_qalora": false,
    "qalora_group_size": 16,
    "layer_replication": null,
    "lora_bias": false,
    "target_parameters": null,
    "use_bdlora": null,
    "arrow_config": null,
    "ensure_weight_tying": false
}`

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

`#!/bin/bash

export REWARD_SILENT=1
export REWARD_DEBUG_FILE="/userhome/reward_debug.jsonl"

export REWARD_DEBUG_SAMPLES=10

export TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_cache
export TRITON_CACHE_DIR=/tmp/triton_cache
export TMPDIR=/tmp
export TEMP=/tmp
export TMP=/tmp

mkdir -p /tmp/torchinductor_cache /tmp/triton_cache
export TORCH_COMPILE_DISABLE=1

CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
MODEL_PATH=/userhome/outputs/qwen3-8b-dora64-REA-GEN-merged
TRAIN_FILE=/userhome/bioprobench_grpo.jsonl
TEST_FILE=/userhome/REA-GEN_test_outputs——QWEN3/grpo_test_64_fixed.jsonl
export WANDB_API_KEY="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"

NGPUS_PER_NODE=8
NNODES=1
TRAIN_ROLLOUT_L0G=/userhome/grpo-gen-new
TRAIN_BATCH_SIZE=128
PPO_MINI_BATCH_SIZE=64
PPO_MICRO_BATCH_SIZE_PER_GPU=2
LOG_PROB_MICRO_BATCH_SIZE_PER_GPU=8
MAX_PROMPT_LENGTH=4096
MAX_RESPONSE_LENGTH=4096
PPO_MAX_TOKEN_LEN_PER_GPU=8192

ACTOR_LR=1e-5
ROLLOUT_TP=1
ROLLOUT_N=4
ROLLOUT_GPU_MEM_UTIL=0.65

REWARD_NUM_WORKERS=2

PROJECT_NAME=verl_bioprobench_grpo1e-5
EXPERIMENT_NAME=qwen3_8b_dora64_grpo_biopro_new
SAVE_FREQ=50
TEST_FREQ=20
TOTAL_EPOCHS=3

DATA=(
    algorithm.adv_estimator=grpo
    algorithm.use_kl_in_reward=False
    data.train_files="['${TRAIN_FILE}']"
    data.val_files="['${TEST_FILE}']"
    data.prompt_key=prompt
    data.return_raw_chat=False
    data.train_batch_size=${TRAIN_BATCH_SIZE}
    data.max_prompt_length=${MAX_PROMPT_LENGTH}
    data.max_response_length=${MAX_RESPONSE_LENGTH}
    data.filter_overlong_prompts=True
    data.truncation='error'
    data.shuffle=False
    data.seed=2026
    
)

MODEL=(
    actor_rollout_ref.model.path=${MODEL_PATH}
    actor_rollout_ref.model.use_remove_padding=True
    actor_rollout_ref.model.enable_gradient_checkpointing=True
    +actor_rollout_ref.model.override_config="{attn_implementation:sdpa}"
    actor_rollout_ref.model.lora_rank=64
    actor_rollout_ref.model.lora_alpha=128
    actor_rollout_ref.model.target_modules=all-linear
    actor_rollout_ref.model.lora.merge=True
)

ACTOR=(
    actor_rollout_ref.actor.optim.lr=${ACTOR_LR}
    actor_rollout_ref.actor.ppo_mini_batch_size=${PPO_MINI_BATCH_SIZE}
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=${PPO_MICRO_BATCH_SIZE_PER_GPU}
    actor_rollout_ref.actor.use_dynamic_bsz=True
    actor_rollout_ref.actor.ppo_max_token_len_per_gpu=${PPO_MAX_TOKEN_LEN_PER_GPU}
    actor_rollout_ref.actor.use_kl_loss=True
    actor_rollout_ref.actor.kl_loss_coef=0.001
    actor_rollout_ref.actor.kl_loss_type=low_var_kl
    actor_rollout_ref.actor.fsdp_config.model_dtype=bfloat16
    ++actor_rollout_ref.rollout.load_format=safetensors

)

ROLLOUT=(
    actor_rollout_ref.rollout.name=vllm
    actor_rollout_ref.rollout.tensor_model_parallel_size=${ROLLOUT_TP}
    actor_rollout_ref.rollout.gpu_memory_utilization=${ROLLOUT_GPU_MEM_UTIL}
    actor_rollout_ref.rollout.n=${ROLLOUT_N}
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=${LOG_PROB_MICRO_BATCH_SIZE_PER_GPU}
    actor_rollout_ref.rollout.dtype=bfloat16
    ++actor_rollout_ref.rollout.enforce_eager=True
)

REWARD=(
    reward.num_workers=${REWARD_NUM_WORKERS}
    reward.reward_model.enable=False
    reward.custom_reward_function.path=/userhome/verl/examples/grpo_trainer/reward_with_server.py
    reward.custom_reward_function.name=compute_score
    +reward.custom_reward_function.reward_kwargs.embedding_model_dir=/userhome/all-mpnet-base-v2
)

TRAINER=(
    trainer.balance_batch=True
    trainer.logger='["console","wandb"]'
    trainer.project_name=${PROJECT_NAME}
    trainer.experiment_name=${EXPERIMENT_NAME}
    trainer.n_gpus_per_node=${NGPUS_PER_NODE}
    trainer.nnodes=${NNODES}
    trainer.save_freq=${SAVE_FREQ}
    trainer.test_freq=${TEST_FREQ}
    trainer.total_epochs=${TOTAL_EPOCHS}
    trainer.rollout_data_dir=${TRAIN_ROLLOUT_L0G}
    trainer.validation_data_dir=${TEST_ROLLOUT_L0G}
    +trainer.keep_last_n_checkpoints=3 
    trainer.resume_mode=auto              
    trainer.resume_from_path=null 
    trainer.default_local_dir=/userhome/grpo-gen-new
)

### Expected behavior

i want to get the final lora-model easily and correctly thanks

**Comment by Kirrito-k423** [comment]
Thank you for the detailed bug report with training logs and configuration.

**Analysis of Bug #2 (LoRA rank/alpha reset)**

The root cause is in the `base_model_merger.py` checkpoint saving logic. When `actor_rollout_ref.model.lora.merge=True`, the code path that saves the adapter folder does not preserve the original LoRA configuration. Specifically:

1. The PEFT adapter is saved via `model.save_pretrained(self.config.target_dir, state_dict=state_dict)`
2. However, the `adapter_config.json` is regenerated by PEFT library without respecting the training-time `lora_alpha=128` setting
3. Instead, it defaults to `lora_alpha=0` because the merge operation bypasses the original config

**The fix requires patching `base_model_merger.py` around L306 to preserve config:**

```python
# Before saving, explicitly update the adapter config
adapter_config_path = os.path.join(self.config.target_dir, \"adapter_config.json\")
if os.path.exists(adapter_config_path):
    import json
    with open(adapter_config_path, \"r\") as f:
        config = json.load(f)
    config[\"r\"] = self.lora_rank  # 64 in your case
    config[\"lora_alpha\"] = self.lora_alpha  # 128 in your case
    with open(adapter_config_path, \"w\") as f:
        json.dump(config, f, indent=2)
```

**Regarding Bug #1 (unexpected exit at last step):**

The \"second-to-last seems right\" pattern typically indicates:
1. A checkpoint save failure at the final step (checkpoint post-processing throws an exception)
2. An async worker cleanup race condition when the trainer finishes

Could you share the full training log? Specifically looking for:
- Any error traceback near the end
- Whether `save_freq` aligns with `total_epochs` (e.g., saving at epoch boundary fails)
- The `keep_last_n_checkpoints=3` behavior at final step

Would you like to open a PR with the adapter_config fix? I can help review.

---

感谢您提供的详细错误报告和训练配置。

**Bug #2（LoRA rank/alpha 被重置）分析**

根本原因在于 `base_model_merger.py` 中的 checkpoint 保存逻辑。当 `actor_rollout_ref.model.lora.merge=True` 时，保存 adapter 文件夹的代码路径没有保留原始的 LoRA 配置。具体来说：
1. PEFT adapter 通过 `model.save_pretrained()` 保存
2. 但 `adapter_config.json` 由 PEFT 库重新生成，没有遵循训练时的 `lora_alpha=128` 设置
3. 反而默认使用 `lora_alpha=0`，因为 merge 操作绕过了原始配置

**修复需要在 `base_model_merger.py` L306 附近打补丁，在保存前显式更新 adapter config 的 r 和 lora_alpha 值。**

**关于 Bug #1（最后一步意外退出）：**

\"倒数第二步正常\"的模式通常表明：①最终step的checkpoint保存失败（post-processing抛异常）②异步worker清理时的竞态条件。请提供完整训练日志以便进一步诊断。

---
🤖 This response was generated by an AI assistant (OpenClaw + Minimax 2.7). If you have follow-up questions, please feel free to discuss!
